### PR TITLE
SDL_GetDefaultAudioInfo for coreaudio

### PR DESCRIPTION
**Pull request by jarod-clark**
Added support for SDL_GetDefaultAudioInfo with coreaudio on macOS

## Description
Created a function which is called by SDL_GetDefaultAudioInfo when the audio driver is coreaudio which gets information about the default audio device on macOS.

@flibitijibibo

### CODE REVIEW

File: test/testsurround.c
Comment (flibitijibibo) [inline test/testsurround.c:221]: Minor change to the test logs, don't mind the diff the suggestion shows...
```suggestion
    if (SDL_GetDefaultAudioInfo(&deviceName, &spec, 0)) {
        SDL_Log("Error when calling SDL_GetDefaultAudioInfo: %s\n", SDL_GetError());
    } else {
        SDL_Log("Default audio output info found!\n");
        SDL_Log("Name: %s\n", deviceName);
        SDL_Log("Sampling Rate: %d\n", spec.freq);
        SDL_Log("Number of Channels: %d\n", spec.channels);
        SDL_Log("Audio Format: %d\n", spec.format);
        SDL_free(deviceName);
    }
    
    if (SDL_GetDefaultAudioInfo(&deviceName, &spec, 1)) {
        SDL_Log("Error when calling SDL_GetDefaultAudioInfo: %s\n", SDL_GetError());
    } else {
        SDL_Log("Default audio capture info found!\n");
```

File: src/audio/coreaudio/SDL_coreaudio.m
Comment (flibitijibibo) [inline src/audio/coreaudio/SDL_coreaudio.m:1214]: For this block, are we able to use AudioObjectGetPropertyData with the devid above, but with kAudioDevicePropertyNominalSampleRate? I'm thinking we can reuse this, maybe...

https://github.com/libsdl-org/SDL/blob/0258ce6cc75fe7e33319544d2add3f2bd4c502f6/src/audio/coreaudio/SDL_coreaudio.m#L155-L175

**Review by flibitijibibo — COMMENTED** [review]
Functionally this should be okay on macOS; iOS/tvOS won't work with GetAudioDeviceSpec but I think we could possibly make GetDefaultAudioInfo work with this...?

File: src/audio/coreaudio/SDL_coreaudio.m
Comment (jarod-clark) [reply src/audio/coreaudio/SDL_coreaudio.m:1214]: Yeah. I had started something like that but don't know how to get all the information from coreaudio. According to the documentation in SDL_audio.h, it looks like we need sample frequency, number of channels, and sample format right? I managed to get sample frequency and number of channels, however, I couldn't get sample format. I could revert to what I had and try and figure out how to get sample format?

Also I think I might be misunderstanding part of the code, so for clarification: where does the sample format get stored in SDL_AudioSpec?

Comment (flibitijibibo) [reply src/audio/coreaudio/SDL_coreaudio.m:1214]: Format is a bitmask, so it'll be a premade format like AUDIO_F32.

Luckily the format field is something we're allowed to skip; any field we can't identify can simply be filled with 0, indicating we couldn't figure out that part of the spec (which I think is what we do for GetAudioDeviceSpec anyway).

Comment (jarod-clark) [reply src/audio/coreaudio/SDL_coreaudio.m:1214]: Perfect, I'll roll that change back to what I had then.

File: test/testsurround.c
Comment (sezero) [inline test/testsurround.c:203]: please move these to the top for c90 compatibility.

Comment (sezero) [reply test/testsurround.c:203]: i.e.:
```diff
diff --git a/test/testsurround.c b/test/testsurround.c
index 45c9a58..ff55af7 100644
--- a/test/testsurround.c
+++ b/test/testsurround.c
@@ -135,6 +135,8 @@ fill_buffer(void* unused, Uint8* stream, int len)
 int
 main(int argc, char *argv[])
 {
+    char *deviceName;
+    SDL_AudioSpec spec;
     int i;
 
     /* Enable standard application logging */
@@ -198,10 +200,7 @@ main(int argc, char *argv[])
 
         SDL_CloseAudioDevice(dev);
     }
-    
-    char *deviceName;
-    SDL_AudioSpec spec;
-    
+
     if (SDL_GetDefaultAudioInfo(&deviceName, &spec, 0)) {
         SDL_Log("Error when calling SDL_GetDefaultAudioInfo: %s\n", SDL_GetError());
     } else {
```

Comment (jarod-clark) [reply test/testsurround.c:203]: Done.

Comment (sezero) [reply test/testsurround.c:203]: Can't see any new commits here.

Also: if you sqash your commits into a single one and do a force-push,
the git history will be cleaner.

Comment (jarod-clark) [reply test/testsurround.c:203]: Alright, I'm somewhat new to git so I don't know how to squash commits into a single one. How would one do something like this? (In the meantime I'll be pushing the commits which I forgot to do).

Comment (madebr) [reply test/testsurround.c]: [This small blog post](https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history#squash-commits-together) explains it all well.
Afterwards you need to force push to your git fork (`git push` with `-f` option).

No need to close this pr and re-open a new one because we can also fix this on our side.

File: src/audio/coreaudio/SDL_coreaudio.m
Comment (sezero) [inline src/audio/coreaudio/SDL_coreaudio.m:1197]: A little more move of locals needed (for clang >= 14 as explained in PR/5854)
Also remove some unnecessary initializations and unused locals. Patch attached
here: [patch.txt](https://github.com/libsdl-org/SDL/files/9637413/patch.txt), also inlined below.
```diff
diff --git a/src/audio/coreaudio/SDL_coreaudio.m b/src/audio/coreaudio/SDL_coreaudio.m
index a3ca542..346f180 100644
--- a/src/audio/coreaudio/SDL_coreaudio.m
+++ b/src/audio/coreaudio/SDL_coreaudio.m
@@ -1148,9 +1148,15 @@ COREAUDIO_OpenDevice(_THIS, const char *devname)
 static int
 COREAUDIO_GetDefaultAudioInfo(char **name, SDL_AudioSpec *spec, int iscapture)
 {
-    UInt32 alive = 0;
-    pid_t pid = 0;
-
+    AudioDeviceID devid;
+    AudioBufferList *buflist;
+    OSStatus result;
+    UInt32 size;
+    CFStringRef cfstr;
+    char *devname;
+    int usable;
+    double sampleRate;
+    
     AudioObjectPropertyAddress addr = {
         iscapture ? kAudioHardwarePropertyDefaultInputDevice
                   : kAudioHardwarePropertyDefaultOutputDevice,
@@ -1158,7 +1164,7 @@ COREAUDIO_GetDefaultAudioInfo(char **name, SDL_AudioSpec *spec, int iscapture)
                   : kAudioDevicePropertyScopeOutput,
         kAudioObjectPropertyElementMaster
     };
-        
+    
     AudioObjectPropertyAddress nameaddr = {
         kAudioObjectPropertyName,
         iscapture ? kAudioDevicePropertyScopeInput
@@ -1179,10 +1185,6 @@ COREAUDIO_GetDefaultAudioInfo(char **name, SDL_AudioSpec *spec, int iscapture)
         kAudioObjectPropertyElementMaster
     };
     
-    AudioDeviceID devid;
-    OSStatus result = noErr;
-    UInt32 size = 0;
-    
     // Get the Device ID
     size = sizeof (AudioDeviceID);
     result = AudioObjectGetPropertyData(kAudioObjectSystemObject, &addr,
@@ -1192,9 +1194,8 @@ COREAUDIO_GetDefaultAudioInfo(char **name, SDL_AudioSpec *spec, int iscapture)
         return SDL_SetError("%s: Default Device ID not found", "coreaudio");
     }
     
-    
     // Use the Device ID to get the name
-    CFStringRef cfstr = NULL;
+    cfstr = NULL;
     size = sizeof (CFStringRef);
     result = AudioObjectGetPropertyData(devid, &nameaddr, 0, NULL, &size, &cfstr);
     
@@ -1205,8 +1206,8 @@ COREAUDIO_GetDefaultAudioInfo(char **name, SDL_AudioSpec *spec, int iscapture)
     CFIndex len = CFStringGetMaximumSizeForEncoding(CFStringGetLength(cfstr),
                                                     kCFStringEncodingUTF8);
     
-    char *devname = (char *) SDL_malloc(len + 1);
-    int usable = ((devname != NULL) &&
+    devname = (char *) SDL_malloc(len + 1);
+    usable = ((devname != NULL) &&
               (CFStringGetCString(cfstr, devname, len + 1, kCFStringEncodingUTF8)));
     CFRelease(cfstr);
     
@@ -1231,7 +1232,7 @@ COREAUDIO_GetDefaultAudioInfo(char **name, SDL_AudioSpec *spec, int iscapture)
     // Uses the Device ID to get the spec
     SDL_zero(*spec);
 
-    double sampleRate = 0;
+    sampleRate = 0;
     size = sizeof(sampleRate);
     result = AudioObjectGetPropertyData(devid, &freqaddr, 0, NULL, &size, &sampleRate);
 
@@ -1245,7 +1246,7 @@ COREAUDIO_GetDefaultAudioInfo(char **name, SDL_AudioSpec *spec, int iscapture)
     if (result != noErr)
         return SDL_SetError("%s: Default Device Data Size not found", "coreaudio");
 
-    AudioBufferList *buflist = (AudioBufferList *) SDL_malloc(size);
+    buflist = (AudioBufferList *) SDL_malloc(size);
     if (buflist == NULL)
         return SDL_SetError("%s: Default Device Buffer List not found", "coreaudio");
 
diff --git a/test/testsurround.c b/test/testsurround.c
index 7761817..ff55af7 100644
--- a/test/testsurround.c
+++ b/test/testsurround.c
@@ -138,6 +138,7 @@ main(int argc, char *argv[])
     char *deviceName;
     SDL_AudioSpec spec;
     int i;
+
     /* Enable standard application logging */
     SDL_LogSetPriority(SDL_LOG_CATEGORY_APPLICATION, SDL_LOG_PRIORITY_INFO);
 
```

File: test/testsurround.c
Comment (jarod-clark) [reply test/testsurround.c]: Thanks for the info!

File: src/audio/coreaudio/SDL_coreaudio.m
Comment (sezero) [reply src/audio/coreaudio/SDL_coreaudio.m]: OK, I don't have anything else.

iOS / tvOS build failures still need resolving.

File: test/testsurround.c
Comment (madebr) [reply test/testsurround.c]: I squashed your pr to one commit in [52534bb](https://github.com/libsdl-org/SDL/pull/6277/commits/52534bb800bd726564c26fc3308634756a53dadf)

**Comment by flibitijibibo** [comment]
Looks like the final piece is that iOS/tvOS don't have AudioDeviceID. Do we have an alternative to get the channel count and sample rate there? Or do we have a device format that Apple recommends assuming (i.e. "tvOS wants 5.1 at 48KHz")?

**Comment by icculus** [comment]
Just checking: info.name is meant to be for displaying to the user and not for opening, right? Because on macOS, opening a specific name will turn off the feature where we track the system default device and switch outputs if it changes at runtime.

**Comment by flibitijibibo** [comment]
Yeah, this isn't meant to be used for OpenAudioDevice, NULL is still the best way to open the default. (We should probably add this to the docs, not sure how I missed that...)

**Comment by flibitijibibo** [comment]
Meandered around for iOS/tvOS solutions, @jarod-clark I think something like this will do it for `!MACOSX_COREAUDIO`?

```
AVAudioSession* session = [AVAudioSession sharedInstance];
spec->rate = [session sampleRate];
spec->channels = [session outputNumberOfChannels];
```

**Comment by icculus** [comment]
(the macOS pieces already posted are fine with me, and can be merged once the iOS concerns are resolved.)

**Comment by jarod-clark** [comment]
@slouken I'm looking into building the project for iOS in order to test SDL_GetDefaultAudioInfo. I have a test which I was going to adapt for this purpose (called `testaudioinfo`) which is located in the Xcode project `SDL/Xcode-iOS/test/TestiPhoneOS.xcodeproj`. I'm getting an error (shown below) when I try to build the project inside of Xcode.
Screen Shot 2022-09-27 at 17 24 48
I was told to ask you if you have any insight into how to fix this error?

**Comment by sezero** [comment]
>  in the Xcode project `SDL/Xcode-iOS/test/TestiPhoneOS.xcodeproj`.
> Screen Shot 2022-09-27 at 17 24 48

How many releases ago that hidapi.framework died, but still lives as a
zombie in xcode project files. Oh boy...

Running grep outputs _lots_ of hidapi.framework entries under Xcode-iOS/Test and
Xcode-iOS/Demos. I can remove them with manual editing but someone doing it with
a recent Xcode would be healthier I guess.

**Comment by slouken** [comment]
Yep, the hidapi issue is on my TODO list: https://github.com/libsdl-org/SDL/issues/6212

**Comment by slouken** [comment]
> @slouken I'm looking into building the project for iOS in order to test SDL_GetDefaultAudioInfo. I have a test which I was going to adapt for this purpose (called `testaudioinfo`) which is located in the Xcode project `SDL/Xcode-iOS/test/TestiPhoneOS.xcodeproj`.

In https://github.com/libsdl-org/SDL/commit/229315b014529d48ec7b067741283dedddb1c0b7 I removed the tests from Xcode-iOS and the tests in the Xcode directory have been updated to run on iOS and tvOS. I ran testaudioinfo on the iPhone 14 Pro simulator and it worked well here. Let me know if you run into any issues!

**Comment by flibitijibibo** [comment]
Looks like the rebase went a little crazy, but the changes are at least there for iOS - will run the tests and then I can just squash this down for the final commit on my end.

**Comment by sezero** [comment]
I just rebased it with minor rearrangement for readability.

File: src/audio/coreaudio/SDL_coreaudio.m
Comment (flibitijibibo) [reply src/audio/coreaudio/SDL_coreaudio.m:1159]: Both name and spec aren't initialized, so at minimum:

```
if (name != NULL) {
    *name = NULL; /* Or an allocated string */
}
SDL_zerop(spec);
```

File: src/audio/coreaudio/SDL_coreaudio.m
Comment (sezero) [inline src/audio/coreaudio/SDL_coreaudio.m:1246]: Can't we just change this one to `SDL_zerop(spec);`

File: src/audio/coreaudio/SDL_coreaudio.m
Comment (sezero) [reply src/audio/coreaudio/SDL_coreaudio.m:1159]: The `name` part applies to the macosx case, too.

Comment (jarod-clark) [reply src/audio/coreaudio/SDL_coreaudio.m:1159]: I think the if statement around line 1242 already pretty much does that. Should I change it to `if (name != NULL)` rather than just `if (name)` though?

**Comment by sezero** [comment]
> I just rebased it 

seems like I did it for nothing..

File: src/audio/coreaudio/SDL_coreaudio.m
Comment (jarod-clark) [reply src/audio/coreaudio/SDL_coreaudio.m:1245]: Yeah I'll change it. I just found this line somewhere else and had copied it over.

**Comment by jarod-clark** [comment]
> > I just rebased it
> 
> seems like I did it for nothing..

I apologize for my inexperience with git.

**Comment by sezero** [comment]
> I apologize for my inexperience with git.

_Never_ do new work in the main or master branch: Do it in a new dedicated
branch and create pull requests from _that_ branch, so that you can rebase
to main later easily. (@slouken: should we document this?)

As for this particular PR: Since you are already on the main branch, no
going back from there. But here are some possible pointers - others here
may do things differently but this _should_ be pretty straight-forward:

1. Identify your particular change and create a `diff` for it against
   the mainstream,
2. Then take a fresh clone of mainstream, change its push origin, e.g.:
   `git remote set-url --push origin git@github.com:jarod-clark/SDL.git`
3. Apply the diff you generated and commit it and then push it.

Hope these help a bit.

File: src/audio/coreaudio/SDL_coreaudio.m
Comment (flibitijibibo) [inline src/audio/coreaudio/SDL_coreaudio.m:1247]: We could probably find some clever way to avoid allocating the devname if `name == NULL`, but for now this will patch it up:
```suggestion
    } else {
        SDL_free(devname);
    }
```

File: src/audio/coreaudio/SDL_coreaudio.m
Comment (slouken) [inline src/audio/coreaudio/SDL_coreaudio.m:1247]: Why are we checking name here? Can't we bypass all the work above if name is NULL? Or are we using devname elsewhere?

Comment (jarod-clark) [reply src/audio/coreaudio/SDL_coreaudio.m:1247]: I believe that was a remnant of the old code I had (which used the dev name elsewhere). I'll check it out and fix it.

Comment (flibitijibibo) [reply src/audio/coreaudio/SDL_coreaudio.m:1247]: I don't _think_ it's used elsewhere, so hypothetically yes we could skip the above work if we wanted to. This at least fixes the leak that was here before, so it's still redundant but won't cause stability issues.

File: src/audio/coreaudio/SDL_coreaudio.m
Comment (flibitijibibo) [inline src/audio/coreaudio/SDL_coreaudio.m:1244]:
```suggestion
        }
        *name = devname;
```

**Review by flibitijibibo — APPROVED** [review]
Weird commit log stuff aside this lgtm. @slouken, I can tidy up the logs on my end so once you mash Approve (provided everything is good!) I'll take it from there.

File: test/testaudioinfo.c
Comment (sezero) [inline test/testaudioinfo.c:79]: `MACOSX_COREAUDIO` is internal to SDL coreaudio source and isn't available to test programs: Are you actually seeing any output from this part of the test program on macosx?

File: test/testaudioinfo.c
Comment (sezero) [inline test/testaudioinfo.c:92]: (same as above)

Comment (jarod-clark) [reply test/testaudioinfo.c:92]: Let me fix that (it wasn't executing on iOS/tvOS/iPadOS, but I have a better way to do that now)

File: test/testaudioinfo.c
Comment (sezero) [reply test/testaudioinfo.c:79]: I would suggest initializing `deviceName` at top to NULL, and replace compile time check with a runtime check like `if(deviceName){...}`

File: test/testaudioinfo.c
Comment (sezero) [reply test/testaudioinfo.c:92]: > Let me fix that (it wasn't executing on iOS/tvOS/iPadOS, but I have a better way to do that now)

I suggest this patch: [patch.txt](https://github.com/libsdl-org/SDL/files/9672460/patch.txt) (also inlined below.)
```diff
diff --git a/test/testaudioinfo.c b/test/testaudioinfo.c
index ee22838..c0333b3 100644
--- a/test/testaudioinfo.c
+++ b/test/testaudioinfo.c
@@ -70,35 +70,28 @@ main(int argc, char **argv)
 
     print_devices(0);
     print_devices(1);
-    
-    
+
     if (SDL_GetDefaultAudioInfo(&deviceName, &spec, 0)) {
         SDL_Log("Error when calling SDL_GetDefaultAudioInfo: %s\n", SDL_GetError());
     } else {
         SDL_Log("Default audio output info found!\n");
-#if MACOSX_COREAUDIO // Device name is unavailable for iOS/tvOS
-        SDL_Log("Name: %s\n", deviceName);
+        SDL_Log("Name: %s\n", deviceName != NULL ? deviceName : "unknown");
         SDL_free(deviceName);
-#endif
         SDL_Log("Sampling Rate: %d\n", spec.freq);
         SDL_Log("Number of Channels: %d\n", spec.channels);
         SDL_Log("Audio Format: %d\n", spec.format);
     }
-    
     if (SDL_GetDefaultAudioInfo(&deviceName, &spec, 1)) {
         SDL_Log("Error when calling SDL_GetDefaultAudioInfo: %s\n", SDL_GetError());
     } else {
         SDL_Log("Default audio capture info found!\n");
-#if MACOSX_COREAUDIO // Device name is unavailable for iOS/tvOS
-        SDL_Log("Name: %s\n", deviceName);
+        SDL_Log("Name: %s\n", deviceName != NULL ? deviceName : "unknown");
         SDL_free(deviceName);
-#endif
         SDL_Log("Sampling Rate: %d\n", spec.freq);
         SDL_Log("Number of Channels: %d\n", spec.channels);
         SDL_Log("Audio Format: %d\n", spec.format);
     }
 
-
     SDL_Quit();
     return 0;
 }
diff --git a/test/testsurround.c b/test/testsurround.c
index ff55af7..34b0f6c 100644
--- a/test/testsurround.c
+++ b/test/testsurround.c
@@ -205,18 +205,17 @@ main(int argc, char *argv[])
         SDL_Log("Error when calling SDL_GetDefaultAudioInfo: %s\n", SDL_GetError());
     } else {
         SDL_Log("Default audio output info found!\n");
-        SDL_Log("Name: %s\n", deviceName);
+        SDL_Log("Name: %s\n", deviceName != NULL ? deviceName : "unknown");
         SDL_Log("Sampling Rate: %d\n", spec.freq);
         SDL_Log("Number of Channels: %d\n", spec.channels);
         SDL_Log("Audio Format: %d\n", spec.format);
         SDL_free(deviceName);
     }
-    
     if (SDL_GetDefaultAudioInfo(&deviceName, &spec, 1)) {
         SDL_Log("Error when calling SDL_GetDefaultAudioInfo: %s\n", SDL_GetError());
     } else {
         SDL_Log("Default audio capture info found!\n");
-        SDL_Log("Name: %s\n", deviceName);
+        SDL_Log("Name: %s\n", deviceName != NULL ? deviceName : "unknown");
         SDL_Log("Sampling Rate: %d\n", spec.freq);
         SDL_Log("Number of Channels: %d\n", spec.channels);
         SDL_Log("Audio Format: %d\n", spec.format);
```

**Pull request closed without merging**
